File: csharp_body_deref/Scripts/Characters/AmmoWeapon.cs
// ==================== AoTTG2 cross-reference ====================
// Type: Characters.AmmoWeapon
// Update status: unchanged since the previous dump
// Real method bodies (native C decompile): source/cpp/Characters/AmmoWeapon.c
// Prior real C# source (older reference): Assets/Scripts/Characters/Human/Weapons/AmmoWeapon.cs
// Bodies below are empty IL2CPP metadata stubs; the logic lives in the native C.
// ===============================================================
using Il2CppDummyDll;

namespace Characters;

[Token(Token = "0x2000531")]
internal class AmmoWeapon : ExtendedUseable
{
	[Token(Token = "0x4001832")]
	[FieldOffset(Offset = "0x3C")]
	public int MaxAmmo;

	[Token(Token = "0x4001833")]
	[FieldOffset(Offset = "0x40")]
	public int AmmoLeft;

	[Token(Token = "0x4001834")]
	[FieldOffset(Offset = "0x44")]
	public int MaxRound;

	[Token(Token = "0x4001835")]
	[FieldOffset(Offset = "0x48")]
	public int RoundLeft;

	[Token(Token = "0x6003625")]
	[Address(RVA = "0x41F9900", Offset = "0x41F9900", VA = "0x41F9900")]
	public AmmoWeapon(BaseCharacter owner, int ammo, int round, float cooldown)
	{
	}

	[Token(Token = "0x6003626")]
	[Address(RVA = "0x41FA690", Offset = "0x41FA690", VA = "0x41FA690")]
	public bool NeedRefill()
	{
		return default(bool);
	}

	[Token(Token = "0x6003627")]
	[Address(RVA = "0x41FA6B0", Offset = "0x41FA6B0", VA = "0x41FA6B0", Slot = "4")]
	public override void Reload()
	{
	}

	[Token(Token = "0x6003628")]
	[Address(RVA = "0x41FA6E0", Offset = "0x41FA6E0", VA = "0x41FA6E0", Slot = "5")]
	public override void Reset()
	{
	}

	[Token(Token = "0x6003629")]
	[Address(RVA = "0x41FA700", Offset = "0x41FA700", VA = "0x41FA700", Slot = "8")]
	protected override void OnUse()
	{
	}

	[Token(Token = "0x600362A")]
	[Address(RVA = "0x41FA720", Offset = "0x41FA720", VA = "0x41FA720", Slot = "6")]
	public override bool CanUse()
	{
		return default(bool);
	}

	[Token(Token = "0x600362B")]
	[Address(RVA = "0x41FA750", Offset = "0x41FA750", VA = "0x41FA750", Slot = "7")]
	public override bool HasDurability()
	{
		return default(bool);
	}
}

File: cpp/Characters/AmmoWeapon.c
// Type: Characters.AmmoWeapon
// Ghidra (System V/gcc) decompilation, IL2CPP structs + signatures applied.
// ---- AoTTG2 cross-reference ----
// Update status: unchanged since the previous dump
// C# structure: source/csharp/Scripts/Characters/AmmoWeapon.cs
// Prior real C# source (older reference): Assets/Scripts/Characters/Human/Weapons/AmmoWeapon.cs
// --------------------------------

// Characters.AmmoWeapon$$.ctor
// il2cpp: void Characters_AmmoWeapon___ctor (Characters_AmmoWeapon_o* __this, Characters_BaseCharacter_o* owner, int32_t ammo, int32_t round, float cooldown, const MethodInfo* method);
// 0x42f9900

void Characters_AmmoWeapon___ctor
               (Characters_AmmoWeapon_o *__this,Characters_BaseCharacter_o *owner,int32_t ammo,int32_t round,
               float cooldown,MethodInfo *method)

{
  int iVar1;
  
  iVar1 = 0;
  Characters_ExtendedUseable___ctor((Characters_ExtendedUseable_o *)__this,owner,(MethodInfo *)0x0);
  (__this->fields).AmmoLeft = ammo;
  (__this->fields).MaxAmmo = ammo;
  (__this->fields).RoundLeft = round;
  (__this->fields).MaxRound = round;
  (__this->fields).Cooldown = cooldown;
  if (round == -1) {
    (__this->fields).RoundLeft = ammo;
  }
  else {
    iVar1 = ammo - round;
  }
  (__this->fields).AmmoLeft = iVar1;
  return;
}


// Characters.AmmoWeapon$$NeedRefill
// il2cpp: bool Characters_AmmoWeapon__NeedRefill (Characters_AmmoWeapon_o* __this, const MethodInfo* method);
// 0x42fa690

bool_conflict Characters_AmmoWeapon__NeedRefill(Characters_AmmoWeapon_o *__this,MethodInfo *method)

{
  int iVar1;
  bool_conflict in_EAX;
  int iVar2;
  
  iVar1 = (__this->fields).MaxRound;
  iVar2 = (__this->fields).RoundLeft;
  if (iVar1 != -1) {
    in_EAX = CONCAT31((int3)((uint)in_EAX >> 8),1);
    if (iVar2 < iVar1) {
      return in_EAX;
    }
    iVar2 = iVar2 + (__this->fields).AmmoLeft;
  }
  return CONCAT31((int3)((uint)in_EAX >> 8),iVar2 < (__this->fields).MaxAmmo);
}


// Characters.AmmoWeapon$$Reload
// il2cpp: void Characters_AmmoWeapon__Reload (Characters_AmmoWeapon_o* __this, const MethodInfo* method);
// 0x42fa6b0

void Characters_AmmoWeapon__Reload(Characters_AmmoWeapon_o *__this,MethodInfo *method)

{
  int iVar1;
  int iVar2;
  int iVar3;
  int iVar4;
  
  iVar3 = (__this->fields).MaxRound;
  if (iVar3 != -1) {
    iVar1 = (__this->fields).AmmoLeft;
    iVar2 = (__this->fields).RoundLeft;
    iVar3 = iVar3 - iVar2;
    if (iVar1 < iVar3) {
      iVar3 = iVar1;
    }
    iVar4 = 0;
    if (-1 < iVar3) {
      iVar4 = iVar3;
    }
    (__this->fields).AmmoLeft = iVar1 - iVar4;
    (__this->fields).RoundLeft = iVar4 + iVar2;
  }
  return;
}


// Characters.AmmoWeapon$$Reset
// il2cpp: void Characters_AmmoWeapon__Reset (Characters_AmmoWeapon_o* __this, const MethodInfo* method);
// 0x42fa6e0

void Characters_AmmoWeapon__Reset(Characters_AmmoWeapon_o *__this,MethodInfo *method)

{
  int iVar1;
  int iVar2;
  int iVar3;
  
  iVar1 = (__this->fields).MaxAmmo;
  iVar2 = (__this->fields).MaxRound;
  iVar3 = iVar1 - iVar2;
  if (iVar2 == -1) {
    iVar2 = iVar1;
    iVar3 = 0;
  }
  (__this->fields).AmmoLeft = iVar3;
  (__this->fields).RoundLeft = iVar2;
  return;
}


// Characters.AmmoWeapon$$OnUse
// il2cpp: void Characters_AmmoWeapon__OnUse (Characters_AmmoWeapon_o* __this, const MethodInfo* method);
// 0x42fa700

void Characters_AmmoWeapon__OnUse(Characters_AmmoWeapon_o *__this,MethodInfo *method)

{
  int iVar1;
  
  Characters_BaseUseable__OnUse((Characters_BaseUseable_o *)__this,(MethodInfo *)0x0);
  iVar1 = (__this->fields).RoundLeft;
  if (-1 < iVar1) {
    (__this->fields).RoundLeft = iVar1 + -1;
  }
  return;
}


// Characters.AmmoWeapon$$CanUse
// il2cpp: bool Characters_AmmoWeapon__CanUse (Characters_AmmoWeapon_o* __this, const MethodInfo* method);
// 0x42fa720

bool_conflict Characters_AmmoWeapon__CanUse(Characters_AmmoWeapon_o *__this,MethodInfo *method)

{
  int iVar1;
  bool_conflict bVar2;
  
  bVar2 = Characters_BaseUseable__CanUse((Characters_BaseUseable_o *)__this,(MethodInfo *)0x0);
  if ((char)bVar2 != '\0') {
    iVar1 = (__this->fields).RoundLeft;
    return CONCAT31((int3)((uint)iVar1 >> 8),iVar1 == -1 || 0 < iVar1);
  }
  return 0;
}


// Characters.AmmoWeapon$$HasDurability
// il2cpp: bool Characters_AmmoWeapon__HasDurability (Characters_AmmoWeapon_o* __this, const MethodInfo* method);
// 0x42fa750

bool_conflict Characters_AmmoWeapon__HasDurability(Characters_AmmoWeapon_o *__this,MethodInfo *method)

{
  int iVar1;
  
  iVar1 = (__this->fields).RoundLeft;
  return CONCAT31((int3)((uint)iVar1 >> 8),iVar1 == -1 || 0 < iVar1);
}


